### PR TITLE
BTreeMap: Support custom allocators

### PR DESCRIPTION
**Pull request by TennyZhuang**
I take over the task from an inactive PR #77438, and most of the work is done by the original author @exrook, thanks! If you are back, feel free to checkout the commits here and close my PR.

---

Related:
https://github.com/rust-lang/wg-allocators/issues/7
https://github.com/TimDiekmann/alloc-wg

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -5410,9 +5410,9 @@ dependencies = [
...
 [[package]]
 name = "toml_edit"
-version = "0.14.3"
+version = "0.14.4"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "ba98375fd631b83696f87c64e4ed8e29e6a1f3404d6aed95fa95163bad38e705"
+checksum = "5376256e44f2443f8896ac012507c19a012df0fe8758b55246ae51a2279db51f"
 dependencies = [
  "combine",
  "indexmap",
```
Comment (Amanieu) [inline Cargo.lock:5418]: These changes to Cargo.lock seem to be unintentional, could you revert them?

Comment (TennyZhuang) [reply Cargo.lock:5418]: It always be updated automatically when I do any x.py operations. 😄

**Comment by TennyZhuang** [comment]
@Amanieu can you help to reopen and review #77438 ? The original author is back and will take over it, I’d like to close the PR.

**Pull request closed without merging**